### DISCUSSION TITLE
CSPropConductingSheet: unit for thickness and PCB stackup models

### DISCUSSION THREAD

**Discussion by oberstet** [thliebig/openEMS-Project#160] [Q&A Usage]
Hi all,

I made some progress in my antenna modeling/openEMS adventures, but at this stage now do have some small questions related to **CSPropConductingSheet** important for me as a RF newbie, and any help/tips/hints would be highly appreciated!

sorry for long post, but I wanted to give proper/complete context.

Cheers,
/Tobias

## intro

Related to https://github.com/thliebig/openEMS-Project/discussions/152, I am following my plan of designing a series of antenna reference models for **446 MHz / 400-1000 MHz frequency of interest** hoping this turns out to be a robust design approach that yields a working FINAL antenna.

Thanks to the fantastic hints/tips and help here on the forum, I now do already have the following flavors working

1. dipole (wire based)
2. monopole (wire based)
3. planar monopole (PCB traces)

and I am now moving to create

4. planar helix monopole (PCB traces)
5. FINAL planar antenna (fractal / switched)

and I also still need to work on fully implementing "thirds rule" (both for the wire based and the planar flavors).

> sidenote: I originally _also_ wanted a reference model for a wire based helix as well, but since openEMS/FDTD doesn't support non-uniform meshing, the number of cells explodes, and I actually don't require any wired antenna, so I skipped this track for now

I have been following what Volker suggested in this comment https://github.com/thliebig/openEMS-Project/discussions/152#discussioncomment-7742922 

> One more detail that will be relevant to your work: For PCB stuff where we have lines that are very thin compared to the xy dimensions, it is also possible (and common!) to simulate the metal as zero thickness sheet, and not resolve the 17µm metal thickness at all. The losses are then handled by openEMS with a surface impedance mapped onto the thin sheet, you will find this in the openEMS transmission line examples. So for your meander antenna work, your mesh-from-edges code can be limited to xy plane, and use a single mesh line for teh conductor in z direction.

## questions

**I am posting model and results further below, but my actual questions are:**

1. Is the following true?

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/1df9d35e-95f8-4b99-ae46-37c57d9c6b57)

-----

2. I am modeling my PCB stackup in a _simplified way_, does that make sense / is it a valid approach?

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/0c4e5e2f-931b-4813-8f03-9ea04f2837c0)

The actual stackup I want to use is [Aisler 6-Layer HD PCB Stackup](https://community.aisler.net/t/6-layer-hd-pcb-stackup/1703)

-----

3. In above stackup model, the PCB traces are a single conductive sheet with copper as material - however, I want to use ENIG plating, that is gold-on-copper - should I model the gold plating _and_ the copper as 2 sheets?
I've looked up "skin depth" at 1GHz, and its in µm's, so I would expect the waves going into both gold/copper, but the wavelength is much bigger than metal thickness - should I use 1 sheet with kappa/epsilon average from gold/copper?

## planar monopole results

![Bildschirmfoto vom 2023-12-12 10-44-17](https://github.com/thliebig/openEMS-Project/assets/233340/aa549472-cbb2-4912-a457-68206a5fc73a)

![Bildschirmfoto vom 2023-12-12 10-46-55](https://github.com/thliebig/openEMS-Project/assets/233340/4e9325ff-231d-422a-b127-d66d7aca872c)

![Bildschirmfoto vom 2023-12-12 10-59-26](https://github.com/thliebig/openEMS-Project/assets/233340/1ff1ee38-2a74-4bd0-a3c2-c015e225c131)

## planar helix monopole WIP

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/f1880e05-7617-4667-a507-96fb5aa986c5)

**Comment by biergaizi** [comment]
> I am modeling my PCB stackup in a simplified way, does that make sense / is it a valid approach?

Likely yes, but it depends on exactly what you're modeling and your PCB's stackup. Check your PCB vendor's websites and datasheets to get the stackup. For example, a 4-layer PCB stackup can be:

1. signal (35 micrometer metal)
2. prepreg (0.1 mm dielectric)
3. ground (35 micrometer metal)
4. core (1.6 mm dielectric)
5. ground (35 micrometer metal)
6. prepreg (0.1 mm dielectric)
7. signal (35 micrometer metal).

You can model this PCB using three different methods.

1. **Complete PCB in an air box with PML boundary**: You can certainly model the complete 4-layer PCB realistically without any simplification, as a complete object floating in an air box, with Perfect Matching Layer as the simulation box's boundary condition. Not recommended, but this method is used by some automated tools. It's also potentially useful to investigate EMI/EMC from the board edge (don't do it unless you really know what you're doing).
10. **Model the 4-layer PCB as a 2-layer PCB**: If you're modeling a transmission line in a multi-layer PCB, you only need to model the objects between the signal layer and its reference planes. Because the ground plane is a perfect electric conductor, anything below the first ground plane has no effect to the simulation (unless other electrical connections exist). So to model a microstrip between layer 1 and layer 2, a simplified model will have two metal layers and one dielectric layer. This is usually a good starting point.
11. **Model the 4-layer PCB as a 1-layer PCB**: To do more simplification, note that the ground plane is already an infinite Perfect Electric Conductor, so you don't need to explicitly model this final layer of metal in your model, instead the boundary condition of the bottom surface of the simulation box can be set to Perfect Electric Conductor, so this ground plane is implicitly modeled. Now you only have two layers, a dielectric layer which is touching the bottom of the simulation box, and a metal layer on top of the dielectric layer. This method is used by many openEMS examples. 

> however, I want to use ENIG plating, that is gold-on-copper - should I model the gold plating and the copper as 2 sheets?

TL;DR: Just model the regular copper. Don't attempt to model ENIG, there's no good solution to do that sofar.

Good question. Unfortunately there's currently no good solution to this problem, hopefully the community will work towards it in 2024.

As a general principle, an inherent problem of FDTD is that it's very difficult (if not impossible) to model very thin layers. Any small features with a much smaller size than a structure's overall size is a no-go, such as a solder mask layer, a ENIG coating layer, or even a microstrip. This is because all electromagnetic field exists in a mesh. To resolve fine structures, you must use a huge number of cells which is usually impractical. For example, in a standard 2-layer board you have a 1600 micrometer substrate and a 35 micrometer trace, a 100x difference.

There are several answers to this problem.

1. **Brute-Force**: Just use an extremely high number of cells. It's only practical for toy examples, not any remotely realistic circuit board designs - but yes, it can model high-frequency behaviors such as skin effect directly from Maxwell's equations. However, extremely high timesteps imply potential numerical instability.

    1.1. **Drude-Lorentz model**:  In openEMS, there's another limitation, `AddMetal()` only supports Perfect Electric Conductor with no resistance, which would prevent you from doing so. But in principle, openEMS supports the Drude-Lorentz model which can model a metal, in fact it's the standard solution (which means even slower simulation, but since we're already talking about an impractical solution, why not...). 

2. **Zero-Thickness Conductor Model**: The standard solution to thin conductor modeling in FDTD is using a fictitious object called a "thin conducting sheet". It has no physical thickness. You still specify a thickness to allow the "fake" mathematical object to model the electrical behaviors of lossy and dispersive conductor. The underlying physics such as electromagnetic fields is *not* modeled, so no, you can't see skin effect directly, but you'll get reasonable values for dispersion and attenuation for practical purposes. In openEMS, it's called `AddConductingSheet()`.

    2.1. **Conducting Sheets Cannot be Overlapped**: Because the thin conductor sheet model is a mathematical model of an entire *complete* conductor and has zero thickness, you can't place anything above it and expect to get correct results. 

So what to do when you really want to model a solder mask layer or a ENIG coating layer? For ENIG, *perhaps* you can "fudge" the value of conductivity of copper to artificially increase the losses. For solder mask, *perhaps* you can put a non-physical layer of dielectric on top of the trace, with the thickness artificially increased while tweaking its dielectric constant to force a "fit". Both probably work for narrow-band simulations.

But practically, no, just ignore them entirely. They're small anyway, especially at 1 GHz. Don't worry about them.

In the future, I hope the community can develop an improved conducting sheet model with a correction factor for ENIG.

**Comment by oberstet** [comment]
fantastic advice! thank you so much=)

I am still digesting, experimenting, will post again once done, and then make a final decision choosing one of the options you hinted at. the one that fits best for my situation, and also my achievable level of know how, means and equipment.

btw, I've also found

https://github.com/toammann/OpenEMS_MultiTermDebyeExperiments/blob/7cd3d29c0005c8293cd9aa2f296592d6e18179b1/OPENEMS_microstripAisler6Lay/microstripAisler6Lay.m#L175

this is quite new, modeling exactly the Aisler stackup I mentioned above. and in fact it also [refers](https://github.com/toammann/OpenEMS_MultiTermDebyeExperiments/tree/7cd3d29c0005c8293cd9aa2f296592d6e18179b1#experiments-with-openems-and-multi-term-debyedjordjevic-sarkar-models) to a discussion including you and Volker;)

I don't really understand the physics/math, can only follow "roughly", I can certainly read the code, and I _could_ port that Matlab code to Python, but it is actually well beyond of what I likely really need and should focus on (which is _antenna_ geometry, size and bandwidth). anyways, will post again later.

**Comment by oberstet** [comment]
ok, got some results .. not yet complete, but still. I have added batch simulation/results for different flavors

https://gist.github.com/oberstet/92dc77b1df90404586e122d83fb8951d

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/ed55546a-9dcc-43d7-bc1f-2573752d1924)

-----

I am implementing the following PCB simulation models

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/49c88cad-b9b7-4681-9c2d-5c49a1ccbfc3)

-----

stuff is going forward I feel;) now, completing above PCB sim models and polishing up everything, like persisting results and allowing to open appcsxcad to view or plot flavors from the batch results etc will take another day.

**
however, I now figure my next main problem will be

how do I decide which of the models is not only "the best" from the mere result data, but the most real, as in, a real PCB will show close results using real instruments?
**

because optimizing each of the flavors for details like impedance or what can be done of course in simulation, but that doesn't mean such an optimized design might show as good as data in the real world, right?

if so, then how do people approach this without manufacturing dozens and dozens of PCB prototypes?

are people calibrating their simulated models using a few prototypes, and then again select from a larger number of (then calibrated) simulated models?

obviously, I would love to just compute the best model, have it manufactured and have it show perfect results identical to my model;)

how are people approaching this fundamental problem?

**Reply by KJ7LNW** [reply]
Were you able to get results with your planar helical antenna?

Is it axial (directional) or normal (ie, like a dipole) ?

**Reply by oberstet** [reply]
> Were you able to get results with your planar helical antenna?

not yet with "correct" meshing, and not yet within my current framework which I now figured out the basics. I have had numbers for helix simulations I did in an earlier phase. I tend to "home in" in a step-wise fashion, throwing away everything again and again, until it converges (I mean code wise).

> Is it axial (directional) or normal (ie, like a dipole) ?

initially, I only thought about planar-monopole. but at this point, why not also try *dipole-monopole*? good point!

or more likely, some kind of *fractal planar dipole* (eg using Dragon curve or what) .. but "fractal" was in my "phase 2" of understanding this stuff where I still thought "size" would be my main problem. anyways, I have code for such already, which I now need to integrate back into the framework, eg it can spill out designs flavors like the following, which I plan to batch simulate and auto-select/optimize then:

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/0089d0eb-040a-4c09-b7c8-c9ae7afd4cca)

---

the fractal curve I selected ^ was _solely_ based on "size": how do I fold a curve into smallest limited space. I now understand that "current" has this property of "direction";) which means, the Hilbert curve likely won't be the best. need to have a different one. anyways, detail, np.

----

rgd antenna types: I initially thought this could be done in 2-3 weeks. turned out wrong of course;) as a total RF newbie, I was missing some crucial elements, but these have in a way settled, and I now pivot back to original goal. anyways, to give you a clue, here is the top level of the framework I am working on to actually design this antenna

```
(cpy312_1) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ rfminer --help
Usage: rfminer [OPTIONS] COMMAND [ARGS]...

Options:
  --unit FLOAT                  Unit of measurement for all lengths in the
                                simulation in meters, defaults to mm.
  --freq_unit FLOAT             Unit of measurement for all frequencies in the
                                simulation in Hz, defaults to MHz.
  --foi TEXT                    Frequencies of interest (priority ordered list
                                of pairs of floats), for example
                                `446.0:446.2,200:1000`.
  --mesh_refinement TEXT        Level of mesh refinement to set up the
                                geometries for and apply in the simulation,
                                defaults to 'standard' (choose from ['coarse',
                                'moderate', 'standard', 'fine', 'dense'])
  --boundary_conditions TEXT    Boundary conditions along 3D simulation space
                                axis, must be exactly 6 values from ['PEC',
                                'PMC', 'MUR', 'PML_8'], using whitespace or
                                ':' or ',' as separator
  --pcb_model_type TEXT         PCB model type to use in simulations of planar
                                antennas (this is _not_ used for regular non-
                                planar antennas). Must be one of ['single-
                                conductive-layer', 'single-conductive-layer-
                                and-substrate', 'two-conductive-layers-and-
                                substrate']
  --length_correction FLOAT     FIXME
  --limit_timesteps INTEGER     Limit number of timesteps simulated (per
                                simulation instance)
  --limit_end_criteria INTEGER  Simulation end criteria in dB, one of [-10,
                                -20, -30, -40, -50, -60]
  --length_correction FLOAT     Use this factor applied to wave length when
                                computing antenna dimensions from wave length
                                (to account for finite reactance)
  --output_dir OPTIONAL         Output directory used for simulation. If none
                                is given, a new directory within the system
                                default temporary files location will be used.
  --debug / --no-debug          Enable debug log output, defaults to `False`.
  --appcsxcad OPTIONAL          Full local filesystem path to AppCSXCAD
                                executable.
  --help                        Show this message and exit.

Commands:
  dipole           Dipole antenna model and simulation commands.
  helix            Helix antenna model and simulation commands.
  monopole         Monopole antenna model and simulation commands.
  planar-fractal   Planar fractal antenna model and simulation commands.
  planar-helix     Planar helix antenna model and simulation commands.
  planar-monopole  Planar monopole antenna model and simulation commands.
  schema           Show RF Miner data model Flatbuffers schema files.
  yagi             Yagi beam antenna model and simulation commands.
(cpy312_1) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ 
```

**Reply by VolkerMuehlhaus** [reply]
> I have added batch simulation/results for different flavors

You can simplify things: there is no need to simulate the same geometry using different port impedances. You could re-normalize S-parameters to other reference impedances at any time. The small differences in your table for calculated input impedance at different port impedance indicates some accuracy error somewhere, in theory it should agree.

I would stick to a single port impedance, the usual value is 50 Ohm to be compatible with VNA measurements. 

> because optimizing each of the flavors for details like impedance or what can be done of course in simulation, but that doesn't mean such an optimized design might show as good as data in the real world, right?

This is two topics in one: 
(a) is the antenna useful
(b) does measurement agree with simulation

(a) The radiation pattern needs to fit the application (we often have small size for PCB-integrated antennas, which tends to radiate more or less omnidirectional -> useful in many cases) and the internal losses in the antenna need to be considered: a lossy antenna can be nicely matched and wideband, but radiate only a small fraction of the input power. This means we need to look at radiation efficiency, which is 1.0 = 100% for a lossless antenna and drop due to losses in the antenna

(b) The key to reliable simulation results is proper use of ports and good meshing, with enough mesh density to resolve the geometry details and resolve fields at the conductors. Of course, material properties need to be correct as well, but for PCB antennas the most important source of loss inside the antenna is the PCB dielectric (loss tangent). Conductor losses are less critical for typical PCB antennas.

> the fractal curve I selected ^ was solely based on "size": how do I fold a curve into smallest limited space. I now understand that "current" has this property of "direction";) which means, the Hilbert curve likely won't be the best. 

Exactly, you don't want current flow where radiated fields would cancel. That would possibly give nice antenna matching, at poor antenna efficieny. I already created such poor designs with folded/meandered Monopoles myself, so use a meander where the main direction is straight on without much back and forth.

~~

Another question: is there any reason why you calculate different ground plane sizes for a given monopole size? The usual workflow is that we start from a given PCB size (defined by the PCB guys to fit all the non-RF stuff) and then tweak the antenna to work with the given PCB. So I assume your workflow is just to get familiar with sweeping some geometry?

**Reply by VolkerMuehlhaus** [reply]
You might find this Texas Instrument overview on PCB antennas useful, which includes links to the different design appnotes.
https://www.ti.com/lit/an/swra351b/swra351b.pdf

**Comment by oberstet** [comment]
thanks again @VolkerMuehlhaus ! also for the TI app note. I will have a deep look.

>  is there any reason why you calculate different ground plane sizes for a given monopole size? 

this is just to gain understanding.

my PCB is already fixed:

* 68x129mm, 6 layer AISLER == groundplane

the antenna requirements are also fixed:

* frequency of interest: 446 & 863 MHz *primary* plus 400-1000 MHz *secondary*
* PCB antenna on above process
* handheld device, omnidirectional antenna, 5W power max
* max 20x76mm antenna (instance) size. maybe a few mm more widths available
* **two** antenna _instances_ L/R at top of above groundplane (thus the complete PCB is 68x129+76 mm .. the space in between the antenna instances is "almost" empty)

the broadband aspect is important for the application, as I want to do SDR based "TV Whitespace" as well

----

still catching up with all your/everyone's comments, tips and help. I am getting there;) will be back.
